service/r2/store/kv/store.go

```diff
@@ -366,28 +373,59 @@ func (s *store) DeleteRollupRule(
 func (s *store) FetchRollupRuleHistory(namespaceID, rollupRuleID string) ([]*models.RollupRuleView, error) {
 	rs, err := s.ruleStore.ReadRuleSet(namespaceID)
 	if err != nil {
-		return nil, s.handleUpstreamError(err)
+		return nil, handleUpstreamError(err)
 	}
 
 	rrs, err := rs.RollupRules()
 	if err != nil {
-		return nil, s.handleUpstreamError(err)
+		return nil, handleUpstreamError(err)
 	}
 
 	hist, exists := rrs[rollupRuleID]
 	if !exists {
-		return nil, s.rollupRuleNotFoundError(namespaceID, rollupRuleID)
+		return nil, rollupRuleNotFoundError(namespaceID, rollupRuleID)
 	}
 	return hist, nil
 }
 
 func (s *store) Close() { s.ruleStore.Close() }
 
-func (s *store) newUpdateMeta(uOpts r2.UpdateOptions) rules.UpdateMetadata {
+func (s *store) UpdateRuleSet(
+	rsChanges changes.RuleSetChanges,
+	version int,
+	uOpts r2store.UpdateOptions,
+) (*models.RuleSet, error) {
+	rs, err := s.FetchRuleSet(rsChanges.Namespace)
+	if err != nil {
+		return nil, handleUpstreamError(err)
+	}
+	if version != rs.Version() {
+		return nil, r2.NewConflictError("mismatched ruleset version number")
+	}
+
+	mutable := rs.ToMutableRuleSet().Clone()
+	err = mutable.ApplyRuleSetChanges(rsChanges, s.newUpdateMeta(uOpts))
```

Comment:
Ah interesting I see, yeah I think you need to set the mutable ruleset version before apply ruleset changes, because that's the ruleset version used for check and set during a store operation in KV. You can add a `SetVersion` API for the `MutableRuleSet` interface and pull that change in here to do that.

Reply:
yeah let me do some digging. I dont think I was doing this in the past and it worked in staging, but I agree it does seem like I should need to update the version explicitly

Reply:
It does not look like any of the other methods update the version by hand so I guess that is taken care of for me. Which means we would need to expose a new KV method that does not increment the version, which does not seem like a pattern we want to support.

Reply:
I more meant we need to add a `SetVersion` method on the `MutableRuleSet` interface to set the version on the in-memory mutable ruleset (not the ruleset stored in kv) because that's the version used for check and set when persisting changes to kv, and then you can use it like below
```
rs, err := s.FetchRuleSet(rsChanges.Namespace)
if err != nil {
  return nil, handleUpstreamError(err)
}
mutable := rs.ToMutableRuleSet().Clone()
mutable.SetVersion(version)
if err = mutable.ApplyRuleSetChanges(rsChanges, s.newUpdateMeta(uOpts)); err != nil {
  return nil, handleUpstreamError(err)
}
if err = s.ruleStore.WriteRuleSet(mutable); err != nil {
  return nil, handleUpstreamError(err)
}
return s.FetchRuleSetSnapshot(rsChanges.Namespace)
```

Reply:
We can sync up next week in person but I think that is problematic because you are fetching a potentially greater version than the one submitted by the user, setting the version to what the user set, applying the changes, so now you have a ruleset that should not really exist and then submitting it. It will fail which is what you want, but your intermediary ruleset is flawed and that seems kind of strange to me. 
But functionally both way of doing it get you to the same desired outcome.

Reply:
Sure we can sync up, btw this is commonly how to make use of check and set based APIs, i.e.,
you fetch the change, apply the change *assuming* the fetched version is the desired version, and commit the change via check and set by checking the underlying version is the desired version. The end result is:
* If the underlying version is indeed the desired version, the change is successfully applied and the underlying version in KV is incremented.
* If the underlying version is changed after the user fetches the change and before the change is committed, the change will fail with a `StaleDataError`, and the user needs to retry with no impact since nothing is persisted.

Reply:
Also yeah I think both ways get to the desired outcome. I'm also fine if you want to keep it the way it is, if so please add some clarifying comments explaining why this would achieve the desired outcome, e.g., something alone the lines of
```
If the current KV version matches the desired version, the persistence operation will succeed if the underlying KV version remains the same before the change is committed and fail otherwise. If the current KV version doesn't match, we will fail fast.
```
since that's not immediately obvious.

Reply:
I guess if I feel I have to leave a github comment on some code explaining how it works that is a good indicator that I should probably have a better in-code comment.